librz/util/subprocess.c

```diff
@@ -562,6 +571,17 @@ RZ_API RzSubprocess *rz_subprocess_start(
 	}
 	proc->stderr_fd = stderr_pipe[0];
 
+	// save current environment and update it with the provided one.
+	// Do this before rz_sys_fork, because after that only async-safe functions
+	// shall be used and setenv is not.
+	HtPP *o_env = ht_pp_new (env_val_dup, env_kv_free, NULL);
+	size_t i;
+	for (i = 0; i < env_size; i++) {
+		char *val = getenv (envvars[i]);
+		ht_pp_insert (o_env, envvars[i], val);
+		rz_sys_setenv (envvars[i], envvals[i]);
+	}
```

Comment:
I don't think we should do this before the fork if possible since then for a short amount of time you will mess up the parent process's env. For example if we later start a debugee with the subprocess api, it would leak the child env into the parent.
Some alternatives:
* Something like https://phabricator.kde.org/R285:bb7c80717a60be7ae00cdf7e05c787217d4403c0 with `execvpe()` though that is a GNU extension but `execle()` is not.
* Maybe manually setting `environ` directly

Reply:
I think setting `environ` manually might be the best solution. For a bit of context, I took a quick look at glibc code of `setenv()`: https://sourceware.org/git/?p=glibc.git;a=blob;f=stdlib/setenv.c;h=c6c6d02fff8d0a172f63869ddb47109cb4f19e63;hb=HEAD#l116 It uses a lock so that would be directly an example where a `setenv()` implementation can deadlock after `fork()`. Basicaly what we would have to do is write something similar as this `__add_to_environ()` but without locking.

Reply:
What about just using `execve`?

Reply:
`execve` takes a raw file path while `execvp*` also take into account the `PATH`. Depends of course on what is desired but I for rz-test at least it would be the latter.